### ISSUE TITLE
[BUG] storageState in globalSetup will has empty cookies in Playwright headless mode

### ISSUE THREAD

**Issue by pawlarius** [microsoft/playwright#15481]
**Context:**
- Playwright Version: @playwright/test version ^1.23.1
- Operating System: macOS Big Sur v11.6 M1 Chip
- Node.js version 16.13.1
- Browser: Chromium

**Code Snippet**

StorageState failed to save cookies in headless mode enabled from Playwright globalSetup.

I define a globalSetup for playwright.config.ts

```
const config: PlaywrightTestConfig = {
  ...  
  globalSetup: require.resolve('./e2e/helpers/globalSetup'),
  ...
}
```

Here's my `e2e/helpers/globalSetup.ts` content:

```
const globalSetup = async (config: FullConfig) => {
  const browser = await chromium.launch();
  const baseURL = config.projects[0].use.baseURL;

  const page = await browser.newPage();
  // do logic to login

  // save session to owner.json file
  await page.context().storageState({ path: 'e2e/mocks/owner.json' });

  await browser.close();
};

```

And in my spec file, I just called them:

```
test.describe.parallel('As owner', () => {
  test.use({ storageState: getStorageState('e2e/mocks/owner.json') });

  test('should be able to see home page', async ({ page }) => {
    await page.goto('/');
    await expect(page.locator('text=HomeIsHere')).toBeVisible();
  });
});

```

This is all working fine if I run the playwright WITHOUT headless mode, I can see my file `e2e/mocks/owner.json` has the cookies stored properly.

But when I enabled headless mode, I can see the save to storageState that is called from globalSetup file will override my 'e2e/mocks/owner.json' file with empty cookies and context like this:

```
{
  "cookies": [],
  "origins": []
}
```

I tried some ways that led me to this conclusion:
- Disable headless mode and run Playwright in debug mode: Cookies is stored
- Remove globalSetup and setup another spec file to store session: Cookies is stored when called with `test.use({ storageState: '' })`

I'm not sure whether someone has found solution to this, I tried checking others reported bug here, but none has a working solution in this case and some are just people keep reporting they have issues but the issue thread is closed :(

- https://github.com/microsoft/playwright/issues/9112
- https://github.com/microsoft/playwright/issues/12884
- https://github.com/microsoft/playwright/issues/717

**Comment by mxschmitt** [comment]
There are multiple reasons why this happens, most likely your website behaves different when a headless browser automates it. You can e.g. enable tracing to find out what exactly is going on (library tab): https://playwright.dev/docs/next/trace-viewer

Also you may need to wait until your are actually logged in, instead of directly trying to save the storageState - imagine that you click, Playwright directly saves the storage state but the actual login is happening in the background after calling storageState.

**Comment by pawlarius** [comment]
Hi @mxschmitt I tried using trace-viewer, but I'm not sure what you're trying to suggest here.
I can't see the cookies from the trace-viewer in the headless mode.

<img width="1273" alt="Screen Shot 2022-07-12 at 13 40 59" src="https://user-images.githubusercontent.com/27823957/178425296-04fbdb23-d67a-45a8-8380-1d0eeebb7af5.png">

I also tried to wait before storing contexts with something like this:

```
await page.waitForTimeout(10000);
await context.storageState({ path: 'e2e/mocks/owner.json' });
```

I'm starting to think the issue lies in the globalSetup from playwright.
No matter how long I set the waitForTimeout, the cookies will never be set by playwright.
But if I run with debug mode PWDEBUG=1, and manually pressing continue until all the steps in globalSetup done, the cookies is magically available for access in my spec file.

**Comment by yury-s** [comment]
> No matter how long I set the waitForTimeout, the cookies will never be set by playwright.
But if I run with debug mode PWDEBUG=1, and manually pressing continue until all the steps in globalSetup done, the cookies is magically available for access in my spec file.

There may also be some timing issue related to the delays between individual actions. When you do it manually the delays are longer which gives the page time to do more stuff. 

> Hi @mxschmitt I tried using trace-viewer, but I'm not sure what you're trying to suggest here.
I can't see the cookies from the trace-viewer in the headless mode.

Since the problem in the globalSetup and test runner only records traces for the actions in the tests you'll need to record it by making corresponding api calls manually like this:

```ts
const globalSetup = async (config: FullConfig) => {
  const browser = await chromium.launch();
  const baseURL = config.projects[0].use.baseURL;

  const context = await browser.newContext({baseURL});
  await context.tracing.start({screenshots: true, snapshots: true, sources: true});
  const page = await context.newPage();
  // do logic to login

  await context.tracing.stop({path: 'trace.zip'});
  // save session to owner.json file
  await context.storageState({ path: 'e2e/mocks/owner.json' });

  await browser.close();
};
```

Record the trace in headless and headed mode and see if there is any differences in the network responses etc.

**Comment by kolpav** [comment]
> No matter how long I set the waitForTimeout, the cookies will never be set by playwright.
But if I run with debug mode PWDEBUG=1, and manually pressing continue until all the steps in globalSetup done, the cookies is magically available for access in my spec file.

I have exactly the same problem. I think headless mode works only with unsecured cookies (not having `httpOnly`, `secure`, `sameSite`)
I also think react lazy routes are not very well supported
Maybe unrelated but in headless mode my tests sometimes see contents of `<noscript>`

**Comment by dgozman** [comment]
@kolpav @pawlarius If you could share the repro with us, including the page url we can access locally, we would be able to debug the issue.

**Comment by dgozman** [comment]
We have to reproduce this issue locally to be able to debug it. If you can provide a full repro, please file a new issue with it, and link to this one.

**Comment by bes** [comment]
@dgozman Is it possible to reopen this issue? I am trying to bring up playwright tests against or webpack web app and using a localhost dev server as well as localhost API proxies. 

The cookies we use during local development are `HttpOnly: FALSE`, `Secure: FALSE`.

If you give me some instructions, maybe I can help figure out what's going on?

I feel kind of sad that the `globalSetup` step is very opaque. Not possible to trace, doesn't show up when using `--debug`, etc. It's one of the most important things a web browser does [signing a user in], and if it fails... I can't debug it?

Thanks!

**Comment by dgozman** [comment]
@bes If you have a repro that we can run locally, please file a new issue and we'll definitely look into it. So far, there is no repro, so we are not able to debug.

> I feel kind of sad that the globalSetup step is very opaque. Not possible to trace, doesn't show up when using --debug, etc. It's one of the most important things a web browser does [signing a user in], and if it fails... I can't debug it?

This is indeed a problem that we are working to fix. It is tracked in multiple issues, for example #14895.

**Comment by BobbyMcWho** [comment]
I'm here from #9112, having the same issue. global-setup creates the storage file but the tests don't actually use it. I ended up having to extend the test module, but it feels dirty

```typescript
// my-test.ts
import { test as base } from '@playwright/test';

export type TestOptions = {
  webappHost: string;
  webappEmail: string;
  webappPassword: string;
};

export const test = base.extend<TestOptions>({
  // Define an option and provide a default value.
  // We can later override it in the config.
  webappHost: [process.env.WEBAPP_HOST || '', { option: true }],
  webappEmail: [process.env.WEBAPP_EMAIL || '', { option: true }],
  webappPassword: [process.env.WEBAPP_PASSWORD || '', { option: true }],

  // Override default "page" fixture.
  page: async ({ page, webappHost, webappEmail, webappPassword}, use) => {
    await page.goto(`https://${webappHost}/login`);
    await page.locator('input[name="email"]').fill(webappEmail);
    await page.locator('input[name="password"]').fill(webappPassword);
    await page.locator('button[type="submit"]').click();

    // Each test will get a "page" that already is logged in
    await use(page);
  },
});
```

**Comment by chirsen** [comment]
I have the same problem with macOS Big Sur v11.6 M1 Chip, use webkit  seems work fine.

```
import { webkit } from 'playwright'

const browser = await webkit.launch()

```

**Comment by tsvetkovamaria** [comment]
Maybe this will be helpful for someone. Check if you cookie has the same domain as the domain you're running your tests against.

**Comment by rfbowen** [comment]
For future weary travelers:

I encountered this problem and came to realize the `request` and `page` objects passed into each test are a bit misleading. Issuing a `POST` with `request` _will not_ populate `page` with your cookies. Attempts to save state off of it will fail / be empty.

Disregard the additional arguments and work from `page` alone. In particular, issue your request with `page.context().request.post`. After that point I was able to navigate and use the cookies successfully, as well as persist them.

Definitely caught me off guard!
